display_list/dl_color.h

```diff
@@ -58,58 +77,99 @@ struct DlColor {
     }
     DlScalar f = getAlphaF();
-    return (argb_ & 0xFF000000) |       //
+    return (argb() & 0xFF000000) |      //
            toC(getRedF() * f) << 16 |   //
            toC(getGreenF() * f) << 8 |  //
            toC(getBlueF() * f);
   }
 
   constexpr DlColor withAlpha(uint8_t alpha) const {  //
-    return DlColor((argb_ & 0x00FFFFFF) | (alpha << 24));
+    return DlColor((argb() & 0x00FFFFFF) | (alpha << 24));
   }
   constexpr DlColor withRed(uint8_t red) const {  //
-    return DlColor((argb_ & 0xFF00FFFF) | (red << 16));
+    return DlColor((argb() & 0xFF00FFFF) | (red << 16));
   }
   constexpr DlColor withGreen(uint8_t green) const {  //
-    return DlColor((argb_ & 0xFFFF00FF) | (green << 8));
+    return DlColor((argb() & 0xFFFF00FF) | (green << 8));
   }
   constexpr DlColor withBlue(uint8_t blue) const {  //
-    return DlColor((argb_ & 0xFFFFFF00) | (blue << 0));
+    return DlColor((argb() & 0xFFFFFF00) | (blue << 0));
+  }
+  constexpr DlColor withAlphaF(float alpha) const {  //
+    return DlColor(alpha, red_, green_, blue_, color_space_);
+  }
+  constexpr DlColor withRedF(float red) const {  //
+    return DlColor(alpha_, red, green_, blue_, color_space_);
+  }
+  constexpr DlColor withGreenF(float green) const {  //
+    return DlColor(alpha_, red_, green, blue_, color_space_);
+  }
+  constexpr DlColor withBlueF(float blue) const {  //
+    return DlColor(alpha_, red_, green_, blue, color_space_);
   }
+  DlColor withColorSpace(DlColorSpace color_space) const;
 
   constexpr DlColor modulateOpacity(DlScalar opacity) const {
     return opacity <= 0   ? withAlpha(0)
            : opacity >= 1 ? *this
                           : withAlpha(round(getAlpha() * opacity));
   }
 
-  constexpr uint32_t argb() const { return argb_; }
+  constexpr uint32_t argb() const {
+    return toC(alpha_) << 24 |  //
+           toC(red_) << 16 |    //
+           toC(green_) << 8 |   //
+           toC(blue_) << 0;
+  }
 
-  bool operator==(DlColor const& other) const { return argb_ == other.argb_; }
-  bool operator!=(DlColor const& other) const { return argb_ != other.argb_; }
-  bool operator==(uint32_t const& other) const { return argb_ == other; }
-  bool operator!=(uint32_t const& other) const { return argb_ != other; }
+  bool isClose(DlColor const& other, DlScalar delta = 1.0f / 256.0f) {
+    return color_space_ == other.color_space_ &&
+           std::abs(alpha_ - other.alpha_) < delta &&
+           std::abs(red_ - other.red_) < delta &&
+           std::abs(green_ - other.green_) < delta &&
+           std::abs(blue_ - other.blue_) < delta;
+  }
+  bool operator==(DlColor const& other) const {
+    return alpha_ == other.alpha_ && red_ == other.red_ &&
+           green_ == other.green_ && blue_ == other.blue_ &&
+           color_space_ == other.color_space_;
+  }
+  bool operator!=(DlColor const& other) const {
+    return !this->operator==(other);
+  }
+  bool operator==(uint32_t const& other) const {
```

Comment:
Do we use this? Is it cases where we compare to SkColor?

Reply:
Yep, it's used.  Just removed it and tried to compile.

Reply:
Probably in comparing to SkColor. I want to do a pass and have SkColor only used from within the skia sub-module, this can probably be deleted then...